### ISSUE TITLE
Simplified Module Lookup Plan

### ISSUE THREAD

**Issue by mariari** [anoma/juvix#893]
### Prologue

We wish to encode a complex module system in terms of simple tools (mainly in terms of Σ, Π, and W). Further this means that the first class behavior that we were baking in can (hoepfully) be broken into a library which can fully encapsulate the ideas. What this gives us is the ability to forgoe complex resolution, and be able to reconsider how qualified names work.

Currently we do the very basic thing, where we traverse the `Context` to decide if it's a literal record or error if it faces anything besides an immediate record. However with these new constraints we can do something a bit fancier.

Namely we wish to the take the following approach

### Plan

1. On Record Declarations, note all fields in the record, and have them in a quickly accessible way (reverse lookup map with priority given to the last definition inputted!).
2.  Sort these into a dependency ordering
3. For any instances, we can then lookup which type it must fulfill, and then run a tupling algorithm on it to get the proper definition that respects the core type theory
   - This should be replaced by proper data types once we have them
4. Transform back into records for backend compilation
    - This is much easier if we have meta-data tracking

This blocks approaches which have anonymous record types, however for now that is an acceptable price to pay.

### Blocked on  
- #892 
- #895

**Comment by mariari** [comment]
### Information Storage Strategy

A storage strategy should be laid out as follows
```lisp
NameSpace A:
  contents : namespace-defs
  openList : openList
  …
  recordAssoc :
    { S#{x,y,z} ⟶ [(name, dependcy-ordering)] … }

;; Example in NameSpace Foo
type point = {x ：int, y : int}

;; Translated into
Foo:
  contents : point-def
  openList : [Implicit Prelude]
  recordAssoc :
    { #S{x,y} ⟶ [(point, [x,y])]}
```

This means that every Namespace/Record in the context will record the struct declarations into it via context pass. Then when we encouter a record creation, we can associate it properly, namely by looking up first in our `recordAssoc` map, then in our opened maps for what it could be.

This misses the mark on a few points
- How do we talk about ordering which differes in different branches of a sum type with the same exact fields?
- Do we track opens in expression well enough to do this? This should be an easy fix
- How do we do something like `Foo.{x = 3, y = 4}` currently we only suport `Foo.({x = 3, y = 4})` for literals like this we should have some kind of reader macro support?

**Comment by mariari** [comment]
After discussion today, it seems the issue with the ΣΠW representation is that it causes the encoding for say a pattern matching to be similar to that of a church numerial, thus a pred function would take `O(n)` time, forcing us to do a conversion back into the original representation.

Thus, we have decided for the time being to add a simple ω records theory into core.

**Comment by cwgoes** [comment]
> Thus, we have decided for the time being to add a simple ω records theory into core.

Is there a write-up of what this records theory is? If not, can we write up something brief?

**Comment by mariari** [comment]
> > Thus, we have decided for the time being to add a simple ω records theory into core.
> 
> Is there a write-up of what this records theory is? If not, can we write up something brief?

There seems to be this commit here

https://github.com/anoma/juvix/commit/0826d6d3a51ff35afc577650bcdf556015cea52c

With this, I can make changes for the website to show off these new forms, and create a new PR that atunes itself to these new forms.
